File: agent-integrations/openclaw/plugin/recall_gate.js
const VALID_MODES = new Set(["off", "adaptive", "always"]);

const explicitMemory = [
  "помнишь", "вспомни", "что ты знаешь обо мне", "из памяти", "в памяти",
  "remember", "recall", "from memory", "what do you know about me",
];
const historical = [
  "раньше", "прошлый", "прошлая", "прошлые", "до этого", "в прошлый раз",
  "мы использовали", "мы делали", "previous", "last time", "used before",
  "we used", "we did", "history",
];
const personal = [
  "мои настройки", "мои предпочтения", "мой профиль", "обо мне", "я предпочитаю",
  "my settings", "my preferences", "my profile", "about me", "i prefer",
];
const project = [
  "наш проект", "нашем проекте", "наша система", "нашей системе", "наш сервер",
  "нашем сервере", "наш репозиторий", "нашем репозитории", "этот проект",
  "этот репо", "в репозитории", "our project", "our system", "our server",
  "our repository", "this project", "this repo", "in the repository",
];
const ambiguous = [
  "та модель", "тот конфиг", "тот сервер", "тот проект", "это пофиксили",
  "там агенты", "как тогда", "как раньше", "that model", "that config",
  "that server", "that project", "the previous one", "as before",
];
const continuations = new Set([
  "продолжай", "продолжи", "дальше", "делай дальше", "continue",
  "continue please", "go on", "keep going",
]);
const shortCommands = new Set([
  "да", "нет", "ок", "окей", "делай", "готово", "спасибо", "yes", "no",
  "ok", "okay", "do it", "thanks", "thank you",
]);
const selfContainedPrefixes = [
  "напиши ", "объясни ", "перечисли ", "создай ", "сгенерируй ", "сравни ",
  "что такое ", "почему ", "как сделать ", "write ", "explain ", "list ",
  "create ", "generate ", "compare ", "what is ", "why ", "how to ",
];

function containsAny(text, needles) {
  return needles.some((needle) => text.includes(needle));
}

export function evaluateRecallGate(
  query,
  { mode = "adaptive", hasLiveContext = null, forceFullRecall = false } = {},
) {
  let normalizedMode = String(mode).trim().toLowerCase();
  if (!VALID_MODES.has(normalizedMode)) normalizedMode = "adaptive";
  const text = String(query || "").toLocaleLowerCase().trim().replace(/\s+/gu, " ");

  if (!text) return { shouldRecall: false, reason: "empty", tier: "none" };
  if (forceFullRecall) return { shouldRecall: true, reason: "explicit_full_recall", tier: "full" };
  if (normalizedMode === "off") return { shouldRecall: false, reason: "mode_off", tier: "none" };
  if (normalizedMode === "always") return { shouldRecall: true, reason: "mode_always", tier: "full" };
  if (containsAny(text, explicitMemory)) return { shouldRecall: true, reason: "explicit_memory", tier: "compact" };
  if (containsAny(text, historical)) return { shouldRecall: true, reason: "historical_reference", tier: "compact" };
  if (containsAny(text, personal)) return { shouldRecall: true, reason: "personal_context", tier: "compact" };
  if (containsAny(text, project)) return { shouldRecall: true, reason: "project_context", tier: "compact" };
  if (containsAny(text, ambiguous)) return { shouldRecall: true, reason: "ambiguous_reference", tier: "compact" };

  const command = text.replace(/[.!?,\s]+$/gu, "");
  if (continuations.has(command)) {
    if (hasLiveContext === false) {
      return { shouldRecall: true, reason: "continuation_with_context", tier: "compact" };
    }
    return { shouldRecall: false, reason: "short_command", tier: "none" };
  }
  if (/^(?:привет(?:ствую)?|здравствуй(?:те)?|доброе\s+(?:утро|день|вечер)|hello|hi|hey|good\s+(?:morning|afternoon|evening))[!,.?\s]*$/iu.test(text)) {
    return { shouldRecall: false, reason: "greeting", tier: "none" };
  }

  const arithmetic = text
    .replace(/^(?:сколько\s+будет|посчитай|вычисли|calculate|what\s+is)\s*/iu, "")
    .trim()
    .replace(/\?$/u, "");
  if (/^[\d\s.,()+*/%\-^=]+$/u.test(arithmetic) && /[+\-*/%^]/u.test(arithmetic)) {
    return { shouldRecall: false, reason: "simple_calculation", tier: "none" };
  }

  const words = text.match(/[\p{L}\p{N}_-]+/gu) || [];
  if (/^(?:переведи|перевод|translate)(?:\s+(?:на|в|to)\s+[\p{L}\p{N}_-]+)?\s*[:\-]?\s+/iu.test(text) && words.length <= 16) {
    return { shouldRecall: false, reason: "single_phrase_translation", tier: "none" };
  }
  if (shortCommands.has(command) || (words.length <= 2 && text.endsWith("!"))) {
    return { shouldRecall: false, reason: "short_command", tier: "none" };
  }
  if (selfContainedPrefixes.some((prefix) => text.startsWith(prefix)) || words.length >= 8) {
    return { shouldRecall: false, reason: "self_contained", tier: "none" };
  }
  return { shouldRecall: true, reason: "conservative_fallback", tier: "compact" };
}

class RecallGateMetrics {
  constructor() {
    this.decisions = new Map();
    this.recalls = 0;
    this.injectedTokens = 0;
    this.latencyMilliseconds = 0;
  }

  recordDecision(decision) {
    const outcome = decision.shouldRecall ? "recall" : "skip";
    const key = `${outcome}:${decision.reason}:${decision.tier}`;
    this.decisions.set(key, (this.decisions.get(key) || 0) + 1);
  }

  recordRecall({ latencyMilliseconds = 0, injectedTokens = 0 } = {}) {
    this.recalls += 1;
    this.latencyMilliseconds += Math.max(0, Number(latencyMilliseconds) || 0);
    this.injectedTokens += Math.max(0, Number(injectedTokens) || 0);
  }

  snapshot() {
    return {
      decisions: Object.fromEntries([...this.decisions.entries()].sort()),
      recalls_total: this.recalls,
      injected_tokens_total: this.injectedTokens,
      recall_latency_seconds_sum: this.latencyMilliseconds / 1000,
    };
  }

  reset() {
    this.decisions.clear();
    this.recalls = 0;
    this.injectedTokens = 0;
    this.latencyMilliseconds = 0;
  }
}

export const recallGateMetrics = new RecallGateMetrics();

export function recallGateMetricsSnapshot() {
  return recallGateMetrics.snapshot();
}

File: agent-integrations/shared/recall_gate.py
"""Deterministic, model-free gate for automatic memory recall."""

from __future__ import annotations

import re
from collections import Counter
from dataclasses import dataclass
from threading import Lock

VALID_RECALL_MODES = frozenset({"off", "adaptive", "always"})

_WORD_RE = re.compile(r"[\w-]+", re.UNICODE)
_GREETING_RE = re.compile(
    r"^(?:привет(?:ствую)?|здравствуй(?:те)?|доброе\s+(?:утро|день|вечер)|"
    r"hello|hi|hey|good\s+(?:morning|afternoon|evening))[!,.?\s]*$",
    re.IGNORECASE,
)
_CALC_PREFIX_RE = re.compile(
    r"^(?:сколько\s+будет|посчитай|вычисли|calculate|what\s+is)\s*", re.IGNORECASE
)
_CALC_BODY_RE = re.compile(r"^[\d\s.,()+*/%\-^=]+$")
_TRANSLATION_RE = re.compile(
    r"^(?:переведи|перевод|translate)(?:\s+(?:на|в|to)\s+[\w-]+)?\s*[:\-]?\s+",
    re.IGNORECASE,
)

_EXPLICIT_MEMORY = (
    "помнишь",
    "вспомни",
    "что ты знаешь обо мне",
    "из памяти",
    "в памяти",
    "remember",
    "recall",
    "from memory",
    "what do you know about me",
)
_HISTORICAL = (
    "раньше",
    "прошлый",
    "прошлая",
    "прошлые",
    "до этого",
    "в прошлый раз",
    "мы использовали",
    "мы делали",
    "previous",
    "last time",
    "used before",
    "we used",
    "we did",
    "history",
)
_PERSONAL = (
    "мои настройки",
    "мои предпочтения",
    "мой профиль",
    "обо мне",
    "я предпочитаю",
    "my settings",
    "my preferences",
    "my profile",
    "about me",
    "i prefer",
)
_PROJECT = (
    "наш проект",
    "нашем проекте",
    "наша система",
    "нашей системе",
    "наш сервер",
    "нашем сервере",
    "наш репозиторий",
    "нашем репозитории",
    "этот проект",
    "этот репо",
    "в репозитории",
    "our project",
    "our system",
    "our server",
    "our repository",
    "this project",
    "this repo",
    "in the repository",
)
_AMBIGUOUS = (
    "та модель",
    "тот конфиг",
    "тот сервер",
    "тот проект",
    "это пофиксили",
    "там агенты",
    "как тогда",
    "как раньше",
    "that model",
    "that config",
    "that server",
    "that project",
    "the previous one",
    "as before",
)
_CONTINUATIONS = frozenset(
    {
        "продолжай",
        "продолжи",
        "дальше",
        "делай дальше",
        "continue",
        "continue please",
        "go on",
        "keep going",
    }
)
_SHORT_COMMANDS = frozenset(
    {
        "да",
        "нет",
        "ок",
        "окей",
        "делай",
        "готово",
        "спасибо",
        "yes",
        "no",
        "ok",
        "okay",
        "do it",
        "thanks",
        "thank you",
    }
)
_SELF_CONTAINED_PREFIXES = (
    "напиши ",
    "объясни ",
    "перечисли ",
    "создай ",
    "сгенерируй ",
    "сравни ",
    "что такое ",
    "почему ",
    "как сделать ",
    "write ",
    "explain ",
    "list ",
    "create ",
    "generate ",
    "compare ",
    "what is ",
    "why ",
    "how to ",
)


@dataclass(frozen=True, slots=True)
class RecallGateDecision:
    """One inspectable automatic-recall decision."""

    should_recall: bool
    reason: str
    tier: str = "none"


class RecallGateMetrics:
    """Bounded in-process counters for host diagnostics and tests."""

    def __init__(self) -> None:
        self._lock = Lock()
        self._decisions: Counter[tuple[str, str, str]] = Counter()
        self._recalls = 0
        self._injected_tokens = 0
        self._latency_seconds = 0.0

    def record_decision(self, decision: RecallGateDecision) -> None:
        outcome = "recall" if decision.should_recall else "skip"
        with self._lock:
            self._decisions[(outcome, decision.reason, decision.tier)] += 1

    def record_recall(self, *, latency_seconds: float, injected_tokens: int) -> None:
        with self._lock:
            self._recalls += 1
            self._latency_seconds += max(0.0, latency_seconds)
            self._injected_tokens += max(0, injected_tokens)

    def snapshot(self) -> dict[str, object]:
        """Return a text-free snapshot suitable for a host metrics bridge."""
        with self._lock:
            decisions = {
                f"{outcome}:{reason}:{tier}": count
                for (outcome, reason, tier), count in sorted(self._decisions.items())
            }
            return {
                "decisions": decisions,
                "recalls_total": self._recalls,
                "injected_tokens_total": self._injected_tokens,
                "recall_latency_seconds_sum": self._latency_seconds,
            }


def evaluate_recall_gate(
    query: str,
    *,
    mode: str = "adaptive",
    has_live_context: bool | None = None,
    force_full_recall: bool = False,
) -> RecallGateDecision:
    """Classify a turn without a model, network call, or stored prompt text."""
    normalized_mode = mode.strip().lower()
    if normalized_mode not in VALID_RECALL_MODES:
        normalized_mode = "adaptive"
    text = " ".join(query.casefold().strip().split())

    if not text:
        return RecallGateDecision(False, "empty")
    if force_full_recall:
        return RecallGateDecision(True, "explicit_full_recall", "full")
    if normalized_mode == "off":
        return RecallGateDecision(False, "mode_off")
    if normalized_mode == "always":
        return RecallGateDecision(True, "mode_always", "full")
    if _contains_any(text, _EXPLICIT_MEMORY):
        return RecallGateDecision(True, "explicit_memory", "compact")
    if _contains_any(text, _HISTORICAL):
        return RecallGateDecision(True, "historical_reference", "compact")
    if _contains_any(text, _PERSONAL):
        return RecallGateDecision(True, "personal_context", "compact")
    if _contains_any(text, _PROJECT):
        return RecallGateDecision(True, "project_context", "compact")
    if _contains_any(text, _AMBIGUOUS):
        return RecallGateDecision(True, "ambiguous_reference", "compact")

    command = text.rstrip(".!?, ")
    if command in _CONTINUATIONS:
        if has_live_context is False:
            return RecallGateDecision(True, "continuation_with_context", "compact")
        return RecallGateDecision(False, "short_command")
    if _GREETING_RE.fullmatch(text):
        return RecallGateDecision(False, "greeting")

    arithmetic = _CALC_PREFIX_RE.sub("", text).strip().rstrip("?")
    if arithmetic and _CALC_BODY_RE.fullmatch(arithmetic) and any(
        operator in arithmetic for operator in "+-*/%^"
    ):
        return RecallGateDecision(False, "simple_calculation")

    words = _WORD_RE.findall(text)
    if _TRANSLATION_RE.match(text) and len(words) <= 16:
        return RecallGateDecision(False, "single_phrase_translation")
    if command in _SHORT_COMMANDS or (len(words) <= 2 and text.endswith("!")):
        return RecallGateDecision(False, "short_command")
    if text.startswith(_SELF_CONTAINED_PREFIXES) or len(words) >= 8:
        return RecallGateDecision(False, "self_contained")
    return RecallGateDecision(True, "conservative_fallback", "compact")


def _contains_any(text: str, needles: tuple[str, ...]) -> bool:
    return any(needle in text for needle in needles)
